modules/graph/vertex_map/arrow_vertex_map_impl.h

```diff
@@ -501,30 +551,53 @@ vineyard::Status BasicArrowVertexMapBuilder<OID_T, VID_T>::Build(
       RETURN_ON_ERROR(array_builder.Seal(client, object));
       varray = std::dynamic_pointer_cast<vineyard_oid_array_t>(object);
       this->set_oid_array(fid, label, varray);
-
       // release the reference
       oid_arrays_[label][fid].clear();
     }
     {
-      vineyard::HashmapBuilder<oid_t, vid_t> builder(client);
-      builder.AssociateDataBuffer(varray->GetBuffer());
-
-      auto array = varray->GetArray();
-      vid_t cur_gid = id_parser_.GenerateId(fid, label, 0);
-      int64_t vnum = array->length();
-      builder.reserve(static_cast<size_t>(vnum));
-      for (int64_t k = 0; k < vnum; ++k) {
-        if (!builder.emplace(array->GetView(k), cur_gid)) {
-          LOG(WARNING)
-              << "The vertex '" << array->GetView(k) << "' has been added "
-              << "more than once, please double check your vertices data";
+      // emplace oid -> gid and set o2g
+      if (!std::is_integral<OID_T>::value) {
+        vineyard::HashmapBuilder<oid_t, vid_t> builder(client);
+        builder.AssociateDataBuffer(varray->GetBuffer());
+
+        auto array = varray->GetArray();
+        vid_t cur_gid = id_parser_.GenerateId(fid, label, 0);
+        int64_t vnum = array->length();
+        builder.reserve(static_cast<size_t>(vnum));
+        for (int64_t k = 0; k < vnum; ++k) {
+          if (!builder.emplace(array->GetView(k), cur_gid)) {
+            LOG(WARNING)
+                << "The vertex '" << array->GetView(k) << "' has been added "
+                << "more than once, please double check your vertices data";
+          }
+          ++cur_gid;
         }
-        ++cur_gid;
+        RETURN_ON_ERROR(builder.Seal(client, object));
+        this->set_o2g(
+            fid, label,
+            std::dynamic_pointer_cast<vineyard::Hashmap<oid_t, vid_t>>(object));
+      } else {
+        vineyard::PerfectHashmapBuilder<oid_t, vid_t> builder(client);
+        builder.AssociateDataBuffer(varray->GetBuffer());
+
+        auto array = varray->GetArray();
+        vid_t cur_gid = id_parser_.GenerateId(fid, label, 0);
+        int64_t vnum = array->length();
+        builder.reserve(static_cast<size_t>(vnum));
+        for (int64_t k = 0; k < vnum; ++k) {
+          if (!builder.emplace(array->GetView(k), cur_gid)) {
+            LOG(WARNING)
+                << "The vertex '" << array->GetView(k) << "' has been added "
+                << "more than once, please double check your vertices data";
+          }
+          ++cur_gid;
+        }
+        RETURN_ON_ERROR(builder.Seal(client, object));
+        this->set_o2g_p(
+            fid, label,
+            std::dynamic_pointer_cast<vineyard::PerfectHashmap<oid_t, vid_t>>(
+                object));
       }
-      RETURN_ON_ERROR(builder.Seal(client, object));
-      this->set_o2g(
-          fid, label,
-          std::dynamic_pointer_cast<vineyard::Hashmap<oid_t, vid_t>>(object));
     }
```

Comment:
You need to add a boolean argument to vertex map builder to decide to use perfetch hash or not (as it has lookup performance degeneration).

Reply:
Done